iac-extensions/arm/src/main/java/org/sonar/iac/arm/visitors/ArmSymbolVisitor.java

```diff
@@ -0,0 +1,93 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.arm.visitors;
+
+import java.util.Set;
+import org.sonar.iac.arm.symbols.Symbol;
+import org.sonar.iac.arm.symbols.SymbolTable;
+import org.sonar.iac.arm.symbols.Usage;
+import org.sonar.iac.arm.tree.api.ArmTree;
+import org.sonar.iac.arm.tree.api.File;
+import org.sonar.iac.arm.tree.api.Identifier;
+import org.sonar.iac.arm.tree.api.VariableDeclaration;
+import org.sonar.iac.common.extension.visitors.InputFileContext;
+import org.sonar.iac.common.extension.visitors.TreeVisitor;
+
+/**
+ * Class used to visit a ArmTree and build symbols and their usages for variables.
+ * Those Symbol/Usage can later be used in checks to resolve Argument(really?) or to report issue variable flow.
+ */
+
+public class ArmSymbolVisitor extends TreeVisitor<InputFileContext> {
+
+  private final SymbolTable symbolTable = new SymbolTable();
+
+  private static final Set<ArmTree.Kind> PARENT_KIND_INDICATING_NO_USAGE = Set.of(
+    ArmTree.Kind.VARIABLE_DECLARATION,
+    ArmTree.Kind.OUTPUT_DECLARATION,
+    ArmTree.Kind.PROPERTY);
+
+  public ArmSymbolVisitor() {
+    register(File.class, (ctx, file) -> visitFile(file));
+    register(VariableDeclaration.class, (ctx, variableDeclaration) -> visitVariableDeclaration(variableDeclaration));
+    register(Identifier.class, (ctx, identifier) -> visitIdentifier(identifier));
+  }
+
+  private void visitFile(File file) {
+    file.setSymbolTable(symbolTable);
+  }
+
+  private void visitVariableDeclaration(VariableDeclaration variableDeclaration) {
+    Symbol symbol = symbolTable.addSymbol(variableDeclaration.declaratedName().value());
+    symbol.addUsage(symbolTable, variableDeclaration, Usage.Kind.ASSIGNMENT);
+  }
+
+  private void visitIdentifier(Identifier identifier) {
+    if (!shouldRegisterAsUsage(identifier)) {
+      return;
+    }
+
+    if (identifier.symbol() != null) {
+      // Variable is already visited
+      return;
+    }
+
+    Symbol symbol = symbolTable.getSymbol(identifier.value());
+    if (symbol != null) {
+      symbol.addUsage(symbolTable, identifier, Usage.Kind.ACCESS);
+
+    }
+  }
+
+  private boolean shouldRegisterAsUsage(Identifier identifier) {
+    ArmTree identifierParent = identifier.parent();
+    if (identifierParent == null) {
+      return true;
+    }
+
+    try {
+      identifierParent.getKind();
+    } catch (UnsupportedOperationException e) {
```

Comment:
I don't like that I need to catch UnsupportedOperationException for `Kinds`, maybe we should change something here...
Happens for example in `InterpolatedStringRightPieceImpl`

Reply:
From what I saw we only have 4 elements without kinds: `SyntaxToken` and the three Interpolated string parts.
Doesn't feel right to have those special use case, maybe we should create a `Kind` for them also.
This can probably be tackled in a separate ticket.